# Exercise 4.5

1. Import Libraries/ Data
2. Create a data frame
3. Mixed Data Type
4. FInd Missing Values
5. FInd Duplicates

Task 4.5
2. Check Products Data
3. Check Mixed Data Type in dfords
5. Find ANy missing Values
6. Address missing values: create a copy of the exisitng dataframe
7. Check for duplicates
10. Export
 

In [ ]:
#  Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Import Data
path = r'C:\Users\419182\8-10-2022 Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
dfords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [5]:
dfords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [6]:
# 2 create a data frame
df_test = pd.DataFrame()

In [7]:
# 3 Create a mixed type column 
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# check to see if data has any mixed type values 
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
# convert numeric to string
df_test['mix'] = df_test['mix'].astype('str')

In [11]:
# 4 find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# create dataframe of missing value
df_nan = df_prods[df_prods['product_name'].isnull() ==True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
# compare number of rows in current vs missing 
df_prods.shape

(49693, 5)

In [15]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() ==False]

In [16]:
df_prods_clean.shape

(49677, 5)

In [17]:
# 5 find duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
df_prods_clean.shape

(49677, 5)

In [20]:
# drop duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5

In [22]:
# 2 Check Products Data
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

I would probably investigate the maximum price value.  Based on the products, 99999 would be way too high.  

In [23]:
# 3 Check mixed data type in dfords dataframe 
for col in dfords.columns.tolist() :
    weird = (dfords[[col]].applymap(type) != dfords[[col]].iloc[0].apply
            (type)).any(axis = 1)
    if len (dfords[weird]) > 0:
        print(col)

There are no mixed data types

In [24]:
# 5 Find any missing values
dfords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
evaluation_set                 0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are many missing values (206209) in the days since prior order column.  This makes sense becasue some of the orders could not have had prior orders.  In other words, this could have been the first order. 

In [25]:
# 6 Address missing values: create a copy of the exisitng dataframe
dfords_copy = dfords.copy()                

In [26]:
# Get the columns with missing values
cols_with_missing = (col for col in dfords_copy.columns if dfords[col].isnull().any())

In [27]:
# create new binary column from the column with missing values
for col in cols_with_missing:
    dfords_copy[col + 'prior_order'] = dfords_copy[col].isnull()

In [28]:
# view the new column
dfords_copy

Unnamed: 0  order_id  user_id evaluation_set  order_number  \
0                 0   2539329        1          prior             1   
1                 1   2398795        1          prior             2   
2                 2    473747        1          prior             3   
3                 3   2254736        1          prior             4   
4                 4    431534        1          prior             5   
...             ...       ...      ...            ...           ...   
3421078     3421078   2266710   206209          prior            10   
3421079     3421079   1854736   206209          prior            11   
3421080     3421080    626363   206209          prior            12   
3421081     3421081   2977660   206209          prior            13   
3421082     3421082    272231   206209          train            14   

         order_dow  order_hour_of_day  days_since_prior_order  \
0                2                  8                     NaN   
1                3                  7                    15.0   
2                3                 12                    21.0   
3                4                  7                    29.0   
4                4                 15                    28.0   
...            ...                ...                     ...   
3421078          5                 18                    29.0   
3421079          4                 10                    30.0   
3421080          1                 12                    18.0   
3421081          1                 12                     7.0   
3421082          6                 14                    30.0   

         days_since_prior_orderprior_order  
0                                     True  
1                                    False  
2                                    False  
3                                    False  
4                                    False  
...                                    ...  
3421078                              False  
3421079                              False  
3421080                              False  
3421081                              False  
3421082                              False  

[3421083 rows x 9 columns]

In [29]:
# rename the new column
dfords_copy.rename(columns = {'days_since_prior_orderprior_order':'prior_order'}, inplace = True)

In [36]:
dfords_copy['prior_order'].value_counts()

False    3214874
True      206209
Name: prior_order, dtype: int64

In [30]:
# view new renamed column
dfords_copy

Unnamed: 0  order_id  user_id evaluation_set  order_number  \
0                 0   2539329        1          prior             1   
1                 1   2398795        1          prior             2   
2                 2    473747        1          prior             3   
3                 3   2254736        1          prior             4   
4                 4    431534        1          prior             5   
...             ...       ...      ...            ...           ...   
3421078     3421078   2266710   206209          prior            10   
3421079     3421079   1854736   206209          prior            11   
3421080     3421080    626363   206209          prior            12   
3421081     3421081   2977660   206209          prior            13   
3421082     3421082    272231   206209          train            14   

         order_dow  order_hour_of_day  days_since_prior_order  prior_order  
0                2                  8                     NaN         True  
1                3                  7                    15.0        False  
2                3                 12                    21.0        False  
3                4                  7                    29.0        False  
4                4                 15                    28.0        False  
...            ...                ...                     ...          ...  
3421078          5                 18                    29.0        False  
3421079          4                 10                    30.0        False  
3421080          1                 12                    18.0        False  
3421081          1                 12                     7.0        False  
3421082          6                 14                    30.0        False  

[3421083 rows x 9 columns]

I knew that the missing data actually helped signify whether the order was the first for the customer or if they had prior orders.  As a result of this missing input being important, i chose to create a new column that flagged the data.  If there is a true, in the new prior order column, we know that it is the customers first order.  False means that they did have prior or prior orders. 

In [31]:
# 7 Check for duplicates
dfords_dups = dfords_copy[dfords_copy.duplicated()]

In [32]:
dfords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, evaluation_set, order_number, order_dow, order_hour_of_day, days_since_prior_order, prior_order]
Index: []

In [33]:
dfords_copy.shape

(3421083, 9)

There are no duplicates in the dfords_copy dataframe

In [34]:
# 10 Export
dfords_copy.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))

In [35]:
# 10 Export
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))